File: Ridy-taxi-app/android/common/src/main/java/com/innomalist/taxi/common/networking/socket/interfaces/SocketNetworkDispatcher.kt
package com.innomalist.taxi.common.networking.socket.interfaces

import android.content.Context
import android.util.Log
import com.google.android.gms.maps.model.LatLng
import com.innomalist.taxi.common.Config
import com.innomalist.taxi.common.models.ChatMessage
import com.innomalist.taxi.common.models.Request
import com.innomalist.taxi.common.utils.Adapters
import com.innomalist.taxi.common.utils.AlertDialogBuilder
import com.squareup.moshi.Json
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.json.JSONObject

class SocketNetworkDispatcher : NetworkDispatcher {
    companion object {
        var instance = SocketNetworkDispatcher()
        var currentNamespace: Namespace? = null
    }
    var socket: Socket? = null
    var onNewMessage: ((ChatMessage) -> Unit)? = null
    var onArrived: ((Int) -> Unit)? = null
    var onStarted: ((Request) -> Unit)? = null
    var onTravelInfo: ((LatLng) -> Unit)? = null
    var onNewRequest: ((Request) -> Unit)? = null
    var onFinished: ((FinishResult) -> Unit)? = null
    var onCancel: ((Int) -> Unit)? = null
    var onCancelRequest: ((Int) -> Unit)? = null
    var onDriverAccepted: ((Request) -> Unit)? = null
    var onPaid: ((Int) -> Unit)? = null

    override fun dispatch(event: String, params: Array<Any>?, completionHandler: (RemoteResponse<Any, SocketClientError>) -> Unit) {
        if(socket == null) {
            return
        }
        socket!!.emit(event, params) {
            if ((it.size > 1)) {
                completionHandler(RemoteResponse.createError(SocketClientError.InvalidAckParamCount))
                return@emit
            }
            if ((it.isEmpty())) {
                GlobalScope.launch(Dispatchers.Main) {
                    completionHandler(RemoteResponse.createSuccess(EmptyClass()))
                }
                return@emit
            }
            completionHandler(RemoteResponse.createSuccess(it[0]))
        }
    }

    fun connect(namespace: Namespace, token: String, notificationId: String, completionHandler: (RemoteResponse<Boolean, ConnectionError>) -> Unit) {
        val options = IO.Options()
        currentNamespace = namespace
        options.reconnection = true
        options.query = "token=$token&os=android&ver=60&not=$notificationId"
        socket = IO.socket("${Config.Backend}${namespace.rawValue}", options)
        socket!!.on(Socket.EVENT_CONNECT) {
            completionHandler(RemoteResponse.createSuccess(true))
        }
        socket!!.on(Socket.EVENT_ERROR) {
            //socket!!.disconnect()
            if (it.isEmpty()) {
                completionHandler(RemoteResponse.createError(ConnectionError.ErrorWithoutData))
            } else if (it[0] is JSONObject) {
                Log.e("Error message", (it[0] as JSONObject)["message"] as String)
                completionHandler(RemoteResponse.createError(ConnectionError.TokenVerificationError))
            } else if (it[0] is String) {
                val knownError = ConnectionError(rawValue = it[0] as String)
                if (knownError != null) {
                    completionHandler(RemoteResponse.createError(knownError))
                } else {
                    completionHandler(RemoteResponse.createError(ConnectionError.Unknown))
                }
            } else {
                completionHandler(RemoteResponse.createError(ConnectionError.NotDecodableError))
            }
        }
            socket!!.on("cancelRequest") { item ->
                GlobalScope.launch(Dispatchers.Main) {
                    onCancelRequest?.invoke(item[0] as Int)
                }
            }
        // Driver Events
        socket!!.on("requestReceived") { item ->
            val travel = Adapters.moshi.adapter(Request::class.java).fromJson(item[0].toString())
            GlobalScope.launch(Dispatchers.Main) {
                onNewRequest?.invoke(travel!!)
            }
        }
        socket!!.on("messageReceived") { item ->
            val message = Adapters.moshi.adapter<ChatMessage>(ChatMessage::class.java).fromJson(item[0].toString())
            onNewMessage?.invoke(message!!)
        }
        socket!!.on("cancelTravel") {
            GlobalScope.launch(Dispatchers.Main) {
                onCancel?.invoke(0)
            }
        }
        socket!!.on("paid") {
            GlobalScope.launch(Dispatchers.Main) {
                onPaid?.invoke(0)
            }
        }
        socket!!.on("arrived") {
            GlobalScope.launch(Dispatchers.Main) {
                onArrived?.invoke(0)
            }
        }
        socket!!.on("started") { item ->
            val travel = Adapters.moshi.adapter<Request>(Request::class.java).fromJson(item[0].toString())
            GlobalScope.launch(Dispatchers.Main) {
                onStarted?.invoke(travel!!)
            }
        }
        socket!!.on("travelInfoReceived") { item ->
            val json = item[0] as JSONObject
            val lng = json.getDouble("x")
            val lat = json.getDouble("y")
            val loc = LatLng(lat, lng)
            GlobalScope.launch(Dispatchers.Main) {
                onTravelInfo?.invoke(loc)
            }
        }
        socket!!.on("Finished") { item ->
            GlobalScope.launch(Dispatchers.Main) {
                if(item[1] is Int) {
                    onFinished?.invoke(FinishResult(item[0] as Boolean, (item[1] as Int).toDouble()))
                } else {
                    onFinished?.invoke(FinishResult(item[0] as Boolean, item[1] as Double))

                }
            }
        }
        socket!!.on("driverAccepted") { item ->
            val travel = Adapters.moshi.adapter<Request>(Request::class.java).fromJson(item[0].toString())
            GlobalScope.launch(Dispatchers.Main) {
                onDriverAccepted?.invoke(travel!!)
            }
        }
        socket!!.connect()
    }

    fun disconnect() {
        socket?.disconnect()
    }
}

enum class Namespace(val rawValue: String) {
    Driver("drivers"), Rider("riders")
}

enum class ConnectionError(val rawValue: String) {
    @Json(name = "VersionOutdated")
    VersionOutdated("VersionOutdated"),
    @Json(name="NotFound")
    NotFound("NotFound"),
    @Json(name="NotFound")
    Blocked("Blocked"),
    @Json(name="RegistrationIncomplete")
    RegistrationIncomplete("RegistrationIncomplete"),
    @Json(name="TokenVerificationError")
    TokenVerificationError("TokenVerificationError"),
    @Json(name="NotDecodableError")
    NotDecodableError("NotDecodableError"),
    @Json(name="Unknown")
    Unknown("Unknown"),
    @Json(name="ErrorWithoutData")
    ErrorWithoutData("ErrorWithoutData");

    fun showAlert(context: Context) {
        AlertDialogBuilder.show(context, this.toString(), AlertDialogBuilder.DialogButton.OK, null)
    }

    companion object {
        operator fun invoke(rawValue: String) = values().firstOrNull { it.rawValue == rawValue }

    }
}

data class FinishResult(
        val paid: Boolean,
        val remainingAmount: Double
)

enum class SocketClientError(val rawValue: String) {
    InvalidAckParamCount("InvalidAckParamCount"), RequestTimeout("RequestTimeout");

    companion object {
        operator fun invoke(rawValue: String) = values().firstOrNull { it.rawValue == rawValue }
    }

    val localizedDescription: String
        get() {
            return when (this) {
                InvalidAckParamCount -> "Result parameter count is more than one. It's unexpected."
                RequestTimeout -> "Request Timeout"
            }
        }
}

File: Ridy-taxi-app/ios/common/Networking/WebSocket/Protocols/SocketNetworkDispatcher.swift
//
//  SocketNetworkDispatcher.swift
//
//  Copyright © 2019 Minimal. All rights reserved.
//

import Foundation
import SocketIO
import MapKit

public struct SocketNetworkDispatcher: NetworkDispatcher {
    public func dispatch(event: String, params: [Any]?, completionHandler: @escaping (Result<Any, SocketClientError>) -> Void) {
        if let _params = params {
            socket!.emitWithAck(event, with: _params).timingOut(after: 20) { response in
                if(response.count > 1) {
                    completionHandler(.failure(SocketClientError.InvalidAckParamCount))
                    return
                }
                if(response.count == 0) {
                    completionHandler(.success(try! EmptyClass().asDictionary()))
                    return
                }
                if let str = response[0] as? String, str == "NO ACK" {
                    completionHandler(.failure(.RequestTimeout))
                    return
                }
                completionHandler(.success((response[0])))
            }
        } else {
            socket!.emitWithAck(event).timingOut(after: 15) { response in
                if(response.count > 1) {
                    completionHandler(.failure(SocketClientError.InvalidAckParamCount))
                    return
                }
                if(response.count == 0) {
                    completionHandler(.success(try! EmptyClass().asDictionary()))
                    return
                }
                if let str = response[0] as? String, str == "NO ACK" {
                    completionHandler(.failure(.RequestTimeout))
                    return
                }
                completionHandler(.success((response[0])))
            }
        }
    }
    
    public static var instance = SocketNetworkDispatcher()
    var socket : SocketIOClient?
    var manager: SocketManager?
    public var userType: SocketNamespace?
    
    private init() {
        
    }
    
    public mutating func connect(namespace: SocketNamespace, token: String, notificationId: String, completionHandler: @escaping (Result<Bool, ConnectionError>) -> Void) {
        userType = namespace
        manager = SocketManager(socketURL: URL(string: Config.Backend)!,config:[.connectParams([
            "token" : token,
            "os" : "ios",
            "ver" : Config.Version,
            "not": notificationId
        ])])
        socket = manager!.socket(forNamespace: "/\(namespace.rawValue)")
        socket!.on(clientEvent: .connect) {data, ack in
            completionHandler(.success(true))
        }
        
        socket!.on(clientEvent: .disconnect) {data, ack in
            
        }
        
        socket!.on(clientEvent: .error) { data, ack in
            if data.count < 1 {
                completionHandler(.failure(.ErrorWithoutData))
            } else if let obj = data[0] as? [String: Any] {
                print(obj["message"]!)
                completionHandler(.failure(.TokenVerificationError))
            } else if let string = data[0] as? String {
                let knownError = ConnectionError(rawValue: string)
                if knownError != nil {
                    completionHandler(.failure(knownError!))
                } else {
                    completionHandler(.failure(.Unknown))
                }
            } else {
                completionHandler(.failure(.NotDecodableError))
            }
            
        }
        // Driver Events
        socket!.on("requestReceived") { data, ack in
            let travel = try! Request(from: data[0] as Any)
            NotificationCenter.default.post(name: .requestReceived, object: travel)
        }
        socket!.on("cancelRequest") { data, ack in
            NotificationCenter.default.post(name: .requestCanceled, object: nil)
        }
        socket!.on("messageReceived") { data, ack in
            NotificationCenter.default.post(name: .messageReceived, object: try! ChatMessage(from: data[0]))
        }
        socket!.on("driverInfoChanged") { data, ack in
            UserDefaultsConfig.user = try! (Driver(from: data[0] as Any).asDictionary())
        }
        socket!.on("cancelTravel") { data, ack in
            NotificationCenter.default.post(name: .cancelTravel, object: nil)
        }
        
        socket!.on("paid") { data, ack in
            NotificationCenter.default.post(name: .paid, object: nil)
        }
        
        //Rider Events
        socket!.on("arrived") { data, ack in
            let request = try! Request(from: data[0])
            NotificationCenter.default.post(name: .arrived, object: request)
        }
        socket!.on("started") { data, ack in
            let request = try! Request(from: data[0])
            NotificationCenter.default.post(name: .serviceStarted, object: request)
        }
        socket!.on("cancelTravel") { data, ack in
            NotificationCenter.default.post(name: .serviceCanceled, object: nil)
        }
        socket!.on("riderInfoChanged") { data, ack in
            UserDefaultsConfig.user = try! (Rider(from: data[0] as Any).asDictionary())
        }
        socket!.on("travelInfoReceived") { data, ack in
            let location = try! CLLocationCoordinate2D(from: data[0])
            NotificationCenter.default.post(name: .travelInfoReceived, object: location)
        }
        socket!.on("Finished") { data, ack in
            NotificationCenter.default.post(name: .serviceFinished, object: data)
        }
        socket!.on("driverAccepted") { data, ack in
            let travel = try! Request(from: data[0] as Any)
            NotificationCenter.default.post(name: .newDriverAccepted, object: travel)
        }
        socket!.connect()
    }
    
    public func disconnect() {
        if let s = socket {
            s.disconnect()
        }
    }
}

public enum SocketNamespace: String {
    case Driver = "drivers"
    case Rider = "riders"
}

public enum ConnectionError: String, Error, Codable {
    case VersionOutdated = "VersionOutdated"
    case NotFound = "NotFound"
    case Blocked = "Blocked"
    case RegistrationIncomplete = "RegistrationIncomplete"
    case TokenVerificationError = "TokenVerificationError"
    case NotDecodableError = "NotDecodableError"
    case Unknown = "Unknown"
    case ErrorWithoutData = "ErrorWithoutData"
}

public enum SocketClientError: String, Error, Codable, LocalizedError {
    case InvalidAckParamCount = "InvalidAckParamCount"
    case RequestTimeout = "RequestTimeout"
    
    var localizedDescription: String {
        get {
            switch self {
            case .InvalidAckParamCount:
                return "Result parameter couunt is more than one. It's unexpected."
                
            case .RequestTimeout:
                return "Request Timeout"
            }
        }
    }
}

public extension Notification.Name {
    static let menuClicked = Notification.Name("menuClicked")
    static let socketError = Notification.Name("socketError")
    static let newDriverAccepted = Notification.Name("newDriverAccepted")
    static let arrived = Notification.Name("arrived")
    static let serviceStarted = Notification.Name("serviceStarted")
    static let serviceCanceled = Notification.Name("serviceCanceled")
    static let riderInfoChanged = Notification.Name("riderInfoChanged")
    static let travelInfoReceived = Notification.Name("travelInfoReceived")
    static let serviceFinished = Notification.Name("serviceFinished")
    static let messageReceived = Notification.Name("messageReceived")
    static let driverInfoChanged = Notification.Name("driverInfoChanged")
    static let cancelTravel = Notification.Name("cancelTravel")
    static let paid = Notification.Name("paid")
    static let requestReceived = Notification.Name("requestReceived")
    static let requestCanceled = Notification.Name("requestCanceled")
    static let statusReceived = Notification.Name("statusReceived")
    static let connectedAfterForeground = Notification.Name("connectedAfterForeground")
    static let connectionError = Notification.Name("connectionError")
}
